amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/LeaseCleanupManager.java

```diff
@@ -0,0 +1,299 @@
+                TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Enqueues a lease for deletion.
+     * @param leasePendingDeletion
+     */
+    public void enqueueForDeletion(LeasePendingDeletion leasePendingDeletion) {
+        final Lease lease = leasePendingDeletion.lease();
+        if (lease == null) {
+            log.warn("Cannot enqueue lease {} for deferred deletion - instance doesn't hold the lease for that shard.",
+                    lease.leaseKey());
+        } else {
+            if (!deletionQueue.contains(leasePendingDeletion)) {
+                log.debug("Enqueuing lease {} for deferred deletion.", lease.leaseKey());
+                deletionQueue.add(leasePendingDeletion);
+            } else {
+                log.warn("Lease {} is already pending deletion, not enqueueing for deletion.", lease.leaseKey());
+            }
+        }
+    }
+
+    /**
+     * Returns how many leases are currently waiting in the queue pending deletion.
+     * @return number of leases pending deletion.
+     */
+    public int leasesPendingDeletion() {
+        return deletionQueue.size();
+    }
+
+    private boolean cleanupLease(LeasePendingDeletion leasePendingDeletion) throws InterruptedException, ExecutionException,
+            TimeoutException, DependencyException, ProvisionedThroughputException, InvalidStateException {
+        final Lease lease = leasePendingDeletion.lease();
+        final ShardInfo shardInfo = leasePendingDeletion.shardInfo();
+        final StreamIdentifier streamIdentifier = leasePendingDeletion.streamIdentifier();
+
+        final AWSExceptionManager exceptionManager = createExceptionManager();
+        boolean cleanedUpLease = false;
+        boolean alreadyCheckedForGarbageCollection = false;
+
+        try {
+            if (leasePendingDeletion.pendingForLongerThan(completedLeaseCleanupThresholdMillis)) {
+                if (cleanupLeasesUponShardCompletion) {
+                    Set<String> childShardKeys = leaseCoordinator.leaseRefresher().getLease(lease.leaseKey()).childShardIds();
+                    if (CollectionUtils.isNullOrEmpty(childShardKeys)) {
+                        try {
+                            childShardKeys = getChildShards(shardInfo, streamIdentifier);
+                            updateLeaseWithChildShards(leasePendingDeletion, childShardKeys);
+                        } catch (ExecutionException e) {
+                            throw exceptionManager.apply(e.getCause());
+                        } finally {
+                            alreadyCheckedForGarbageCollection = true;
+                        }
+                    }
+                    cleanedUpLease = cleanupLeaseForCompletedShard(lease, shardInfo, childShardKeys);
+                }
+            }
+
+            if (!alreadyCheckedForGarbageCollection && leasePendingDeletion.pendingForLongerThan(garbageLeaseCleanupThresholdMillis)) {
```

Comment:
Same comment as above for cleanup timers. Basically we need three timers
1. A timer to scan the queue - say 1 min
2. A timer to decide to clean completed shard - say 5 min
3. A timer to clean garbage shards - say 30 min

Now we would run cleanup job 60 times in 1 hour, of which you would have executed completedShard  scan for 12 times and garbageShard scan 2 times.